compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -5,46 +5,125 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "iree/compiler/Codegen/Common/EncodingUtils.h"
+#include "llvm/Support/Debug.h"
 #include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Utils/IndexingUtils.h"
-
+#include "mlir/IR/BuiltinTypes.h"
 namespace mlir::iree_compiler {
 
 using IREE::LinalgExt::EncodingAttr;
 using IREE::LinalgExt::EncodingRole;
 
+// If tensorType has the encoding of a matmul RESULT with narrow N, returns
+// the transposed type. Otherwise, just returns tensorType.
+static RankedTensorType transposeIfNarrowNResult(RankedTensorType tensorType) {
+  EncodingAttr encoding =
+      tensorType.getEncoding().dyn_cast_or_null<EncodingAttr>();
+  if (!encoding || tensorType.getRank() <= 1) {
+    return tensorType;
+  }
+  if (!isNarrowNResult(encoding)) {
+    return tensorType;
+  }
+  if (tensorType.getRank() <= 1) {
+    return tensorType;
+  }
+  auto newRole = encoding.getRole().getValue();
+  TypeAttr originalTypeAttr = encoding.getOriginalType();
+  RankedTensorType originalType;
+  if (originalTypeAttr) {
+    originalType = originalTypeAttr.getValue().dyn_cast<RankedTensorType>();
+  }
+  if (!originalType) {
+    originalType = tensorType;
+  }
+  SmallVector<int64_t> newOriginalShape{originalType.getShape()};
+  auto userIndexingMaps = encoding.getUserIndexingMaps();
+  SmallVector<AffineMap> maps;
+  for (auto a : userIndexingMaps) {
+    maps.push_back(cast<AffineMapAttr>(a).getAffineMap());
+  }
+  auto cDims = linalg::inferContractionDims(maps);
+  SmallVector<int64_t> newShape{tensorType.getShape()};
+  SmallVector<int64_t> permIndices(maps[0].getNumDims());
+  for (int i = 0; i < permIndices.size(); ++i) {
+    permIndices[i] = i;
+  }
```

Comment:
nit: this can be `std::iota(permIndices.begin(), permIndices.end(), 0)` or `auto permIndices = llvm::to_vector(llvm::seq<int64_t>(0, maps[0].getNumDims()));`.